### PR TITLE
P2442R1 Windowing range adaptors: views::chunk and views::slide

### LINKED ISSUE #5271: [2022-02 LWG Motion 12] P2442R1 Windowing range adaptors: views::chunk and views::slide
[P2442R1](https://wg21.link/p2442r1)

### PR DESCRIPTION
**Pull request by jensmaurer**
- Fix return type in description of chunk_view::inner-iterator::operator++.

Fixes #5271
Fixes cplusplus/papers#1108

### CODE REVIEW

File: source/ranges.tex [hunk 1]
```diff
@@ -10040,3 +10070,1603 @@
...
+\begin{codeblock}
+@\exposid{iterator}@<false>(ranges::prev(ranges::end(@\exposid{base_}@), @\exposid{n_}@ - 1, ranges::begin(@\exposid{base_}@)), @\exposid{n_}@)
+\end{codeblock}
+\item
+Otherwise, if \tcode{V} models \libconcept{common_range},
+\begin{codeblock}
+@\exposid{iterator}@<false>(ranges::end(@\exposid{base_}@), ranges::end(@\exposid{base_}@), @\exposid{n_}@)
+\end{codeblock}
+\item
+Otherwise, \tcode{\exposid{sentinel}(ranges::end(\exposid{base_})).}
```
Comment (JohelEGP) [inline source/ranges.tex:11185]:
```suggestion
Otherwise, \tcode{\exposid{sentinel}(ranges::end(\exposid{base_}))}.
```

File: source/ranges.tex [hunk 2]
```diff
@@ -10040,3 +10070,1603 @@
...
+\remarks
+In order to provide the amortized constant-time complexity
+required by the \libconcept{range} concept,
+this function caches the result within the \tcode{slide_view}
+for use on subsequent calls
+when \tcode{V} models \exposconcept{slide-caches-last}.
+\end{itemdescr}
+
+\begin{itemdecl}
+constexpr auto end() const requires slide-caches-nothing<const V>;
```
Comment (JohelEGP) [inline source/ranges.tex:11198]:
```suggestion
constexpr auto end() const requires @\exposconcept{slide-caches-nothing}@<const V>;
```

File: source/ranges.tex [hunk 3]
```diff
@@ -10040,3 +10070,1603 @@
...
+
+\begin{itemdescr}
+\pnum
+\returns
+\tcode{begin() + range_difference_t<const V>(size())}.
+\end{itemdescr}
+
+\begin{itemdecl}
+constexpr auto size() requires sized_range<V>;
+constexpr auto size() const requires sized_range<const V>;
```
Comment (JohelEGP) [inline source/ranges.tex:11209]:
```suggestion
constexpr auto size() requires @\libconcept{sized_range}@<V>;
constexpr auto size() const requires @\libconcept{sized_range}@<const V>;
```

File: source/ranges.tex [hunk 4]
```diff
@@ -10040,3 +10070,1603 @@
...
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\effects
+Equivalent to:
+\begin{codeblock}
+auto sz = ranges::distance(@\exposid{base_}@) - @\exposid{n_}@ + 1;
+if (sz < 0) sz = 0;
```
Comment (JohelEGP) [inline source/ranges.tex:11218]: Looks like `max<`_`T`_`>(0, ...)` for _`T`_ = `decltype(sz)` (current `sz` above).
```suggestion
auto sz = max({}, ranges::distance(@\exposid{base_}@) - @\exposid{n_}@ + 1);
```

File: source/ranges.tex [hunk 5]
```diff
@@ -10040,3 +10070,1603 @@
...
+\begin{codeblock}
+namespace std::ranges {
+  template<@\libconcept{forward_range}@ V>
+    requires @\libconcept{view}@<V>
+  template<bool Const>
+  class slide_view<V>::@\exposid{iterator}@ {
+    using @\exposid{Base}@ = @\exposid{maybe-const}@<Const, V>;                 // \expos
+    iterator_t<@\exposid{Base}@> @\exposid{current_}@   = iterator_t<@\exposid{Base}@>();   // \expos
+    iterator_t<@\exposid{Base}@> @\exposid{last_ele_}@  = iterator_t<@\exposid{Base}@>();   // \expos,
+                                               // present only if \tcode{\exposid{Base}} models \tcode{\exposconcept{slide-caches-first}}
```
Comment (JohelEGP) [inline source/ranges.tex:11235]:
```suggestion
                                               // present only if \exposid{Base} models \exposconcept{slide-caches-first}
```

File: source/ranges.tex [hunk 6]
```diff
@@ -10040,3 +10070,1603 @@
...
+    range_difference_t<@\exposid{Base}@> @\exposid{n_}@ = 0;                    // \expos
+
+    constexpr @\exposid{iterator}@(iterator_t<@\exposid{Base}@> current, range_difference_t<@\exposid{Base}@> n) // \expos
+      requires (!@\exposconcept{slide-caches-first}@<@\exposid{Base}@>);
+
+    constexpr @\exposid{iterator}@(iterator_t<@\exposid{Base}@> current, iterator_t<@\exposid{Base}@> @\exposid{last_ele}@,  // \expos
```
Comment (JohelEGP) [inline source/ranges.tex:11241]:
```suggestion
    constexpr @\exposid{iterator}@(iterator_t<@\exposid{Base}@> current, iterator_t<@\exposid{Base}@> last_ele,  // \expos
```

File: source/ranges.tex [hunk 7]
```diff
@@ -10040,3 +10070,1603 @@
...
+      requires @\exposconcept{slide-caches-first}@<@\exposid{Base}@>;
+  public:
+    using iterator_category = input_iterator_tag;
+    using iterator_concept = @\seebelow@;
+    using value_type = decltype(views::counted(@\exposid{current_}@, @\exposid{n_}@));
+    using difference_type = range_difference_t<@\exposid{Base}@>;
+
+    @\exposid{iterator}@() = default;
+    constexpr @\exposid{iterator}@(@\exposid{iterator}@<!Const> i)
+      requires Const && convertible_to<iterator_t<V>, iterator_t<@\exposid{Base}@>>;
```
Comment (JohelEGP) [inline source/ranges.tex:11252]:
```suggestion
      requires Const && @\libconcept{convertible_to}@<iterator_t<V>, iterator_t<@\exposid{Base}@>>;
```

File: source/ranges.tex [hunk 8]
```diff
@@ -10040,3 +10070,1603 @@
...
+\begin{itemdescr}
+\pnum
+\effects
+Initializes \exposid{current_} with \tcode{current} and
+\exposid{n_} with \tcode{n}.
+\end{itemdescr}
+
+\begin{itemdecl}
+constexpr @\exposid{iterator}@(iterator_t<@\exposid{Base}@> current, iterator_t<@\exposid{Base}@> last_ele,
+              range_difference_t<@\exposid{Base}@> n)
```
Comment (JohelEGP) [inline source/ranges.tex:11326]:
```suggestion
                   range_difference_t<@\exposid{Base}@> n)
```

File: source/ranges.tex [hunk 9]
```diff
@@ -10040,3 +10070,1603 @@
...
+\pnum
+\effects
+Initializes \exposid{current_} with \tcode{current},
+\exposid{last_ele_} with \tcode{last_ele}, and
+\exposid{n_} with \tcode{n}.
+\end{itemdescr}
+
+\begin{itemdecl}
+constexpr @\exposid{iterator}@(@\exposid{iterator}@<!Const> i)
+  requires Const && (convertible_to<iterator_t<V>, iterator_t<@\exposid{Base}@>>;
```
Comment (JohelEGP) [inline source/ranges.tex:11340]:
```suggestion
  requires Const && @\libconcept{convertible_to}@<iterator_t<V>, iterator_t<@\exposid{Base}@>>;
```

File: source/ranges.tex [hunk 10]
```diff
@@ -10040,3 +10070,1603 @@
...
+\exposid{current_} and \exposid{last_ele_} (if present) are
+each equal to \tcode{ranges::next(i)},
+where \tcode{i} is the value of that data member before the call.
+
+\returns
+\tcode{*this}.
+\end{itemdescr}
+
+\begin{itemdecl}
+constexpr iterator operator++(int);
```
Comment (JohelEGP) [inline source/ranges.tex:11384]:
```suggestion
constexpr @\exposid{iterator}@ operator++(int);
```

File: source/ranges.tex [hunk 11]
```diff
@@ -10040,3 +10070,1603 @@
...
+
+\begin{itemdecl}
+constexpr iterator& operator+=(difference_type x)
+  requires @\libconcept{random_access_range}@<@\exposid{Base}@>;
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\expects
+\tcode{\exposid{current_} + x} and \tcode{\exposid{last_ele_} + x} (if present)
```
Comment (JohelEGP) [inline source/ranges.tex:11440]: What does it mean for _`last_ele_`_` + x` to be present?
```suggestion
\tcode{\exposid{current_} + x} and \tcode{\exposid{last_ele_} + x} (if \exposid{last_ele_} is present)
```

File: source/ranges.tex [hunk 12]
```diff
@@ -10040,3 +10070,1603 @@
...
+
+\begin{itemdecl}
+constexpr @\exposid{iterator}@& operator-=(difference_type x)
+  requires @\libconcept{random_access_range}@<@\exposid{Base}@>;
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\expects
+\tcode{\exposid{current_} - x} and \tcode{\exposid{last_ele_} - x} (if present)
```
Comment (JohelEGP) [inline source/ranges.tex:11460]: What does it mean for _`last_ele_`_` - x` to be present?
```suggestion
\tcode{\exposid{current_} - x} and \tcode{\exposid{last_ele_} - x} (if \exposid{last_ele_} is present)
```

File: source/ranges.tex [hunk 13]
```diff
@@ -10040,3 +10070,1603 @@
...
+
+\begin{itemdecl}
+friend constexpr bool operator<(const @\exposid{iterator}@& x, const @\exposid{iterator}@& y)
+  requires @\libconcept{random_access_range}@<@\exposid{Base}@>;
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\returns
+x.\exposid{current_} < y.\exposid{current_}.
```
Comment (JohelEGP) [inline source/ranges.tex:11503]:
```suggestion
\tcode{x.\exposid{current_} < y.\exposid{current_}}.
```

File: source/ranges.tex [hunk 14]
```diff
@@ -10040,3 +10070,1603 @@
...
+\pnum
+\returns
+\tcode{x.\exposid{current_} <=> y.\exposid{current_}}.
+\end{itemdescr}
+
+\begin{itemdecl}
+friend constexpr iterator operator+(const @\exposid{iterator}@& i, difference_type n)
+  requires @\libconcept{random_access_range}@<@\exposid{Base}@>;
+friend constexpr iterator operator+(difference_type n, const @\exposid{iterator}@& i)
+  requires @\libconcept{random_access_range}@<@\exposid{Base}@>;
```
Comment (JohelEGP) [inline source/ranges.tex:11555]:
```suggestion
friend constexpr @\exposid{iterator}@ operator+(const @\exposid{iterator}@& i, difference_type n)
  requires @\libconcept{random_access_range}@<@\exposid{Base}@>;
friend constexpr @\exposid{iterator}@ operator+(difference_type n, const @\exposid{iterator}@& i)
  requires @\libconcept{random_access_range}@<@\exposid{Base}@>;
```

File: source/ranges.tex [hunk 15]
```diff
@@ -10040,3 +10070,1603 @@
...
+
+\begin{itemdescr}
+\pnum
+\returns
+If \exposid{last_ele_} is present,
+\tcode{x.\exposid{last_ele_} - y.\exposid{last_ele_}};
+otherwise, \tcode{x.\exposid{current_} - y.\exposid{cur\-rent_}}.
+\end{itemdescr}
+
+\rSec3[range.slide.sentinel]{Class slide_view::sentinel}
```
Comment (JohelEGP) [inline source/ranges.tex:11598]:
```suggestion
\rSec3[range.slide.sentinel]{Class \tcode{slide_view::\exposid{sentinel}}}
```

File: source/ranges.tex [hunk 16]
```diff
@@ -10040,3 +10070,1603 @@
...
+      operator-(const @\exposid{sentinel}@& y, const @\exposid{iterator}@<false>& x)
+        requires @\libconcept{sized_sentinel_for}@<sentinel_t<V>, iterator_t<V>>;
+  };
+}
+\end{codeblock}
+
+\pnum
+\begin{note}
+\exposid{sentinel} is used
+only when \exposconcept{slide-caches-first}<V> is \tcode{true}.
```
Comment (JohelEGP) [inline source/ranges.tex:11627]:
```suggestion
only when \tcode{\exposconcept{slide-caches-first}<V>} is \tcode{true}.
```

File: source/ranges.tex [hunk 17]
```diff
@@ -10040,3 +10070,1603 @@
...
+\end{note}
+
+\begin{itemdecl}
+constexpr explicit sentinel(sentinel_t<V> end);
```
Comment (JohelEGP) [inline source/ranges.tex:11631]:
```suggestion
constexpr explicit @\exposid{sentinel}@(sentinel_t<V> end);
```

File: source/ranges.tex [hunk 18]
```diff
@@ -10040,3 +10070,1603 @@
...
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\effects
+Initializes \exposid{end_} with \tcode{end}.
+\end{itemdescr}
+
+\begin{itemdecl}
+friend constexpr bool operator==(const iterator<false>& x, const sentinel& y);
```
Comment (JohelEGP) [inline source/ranges.tex:11641]:
```suggestion
friend constexpr bool operator==(const @\exposid{iterator}@<false>& x, const @\exposid{sentinel}@& y);
```

File: source/ranges.tex [hunk 19]
```diff
@@ -10040,3 +10070,1603 @@
...
+\begin{itemdescr}
+\pnum
+\returns
+\tcode{x.\exposid{last_ele_} - y.\exposid{end_}}.
+\end{itemdescr}
+
+\begin{itemdecl}
+friend constexpr range_difference_t<V>
+  operator-(const @\exposid{sentinel}@& y, const @\exposid{iterator}@<false>& x)
+    requires \tcode{sized_sentinel_for}<sentinel_t<V>, iterator_t<V>>;
```
Comment (JohelEGP) [inline source/ranges.tex:11665]:
```suggestion
    requires @\libconcept{sized_sentinel_for}@<sentinel_t<V>, iterator_t<V>>;
```

**Review by JohelEGP — CHANGES_REQUESTED** [review]
`slide_view`'s `begin` and `end` have descriptions like:

_Returns:_
- If _cond_, `a`.
- Otherwise, `b`.

IIRC, they are done like this in `<algorithm>`:

_Returns:_
- If _cond_, returns `a`.
- Otherwise, returns `b`.

The latter is certainly clearer. Is this worth pursuing (at this time or otherwise)?

Also applies to https://github.com/cplusplus/draft/pull/5249/files#diff-015716b0f124dba6acb7cdf24285de5d954d0958ca8abbb2ad82a015d1852a9eR11489-R11493 and https://github.com/cplusplus/draft/pull/5249/files#diff-015716b0f124dba6acb7cdf24285de5d954d0958ca8abbb2ad82a015d1852a9eR11592-R11595.

File: source/ranges.tex [hunk 20]
```diff
@@ -10040,3 +10070,1600 @@
...
+  cout << '[';
+  auto sep = "";
+  for(auto i : r) {
+    cout << sep << i;
+    sep = ", ";
+  }
+  cout << "] ";
+}
+// prints: [1, 2] [3, 4] [5]
+\end{codeblock}
```
[the remark this answers is not in the document]
Comment (jensmaurer) [reply source/ranges.tex:10107]: I think "the above" for this case is good enough.

File: source/ranges.tex (hunk 4 above)
Comment (jensmaurer) [reply source/ranges.tex:11214]: Not doing this for now; please file a separate editorial issue.

File: source/ranges.tex (hunk 5 above)
Comment (jensmaurer) [reply source/ranges.tex:11231]: The font isn't quiite right otherwise, I think.

**Comment by jensmaurer** [comment]
Thanks for all the fixes!

File: source/ranges.tex (hunk 5 above)
Comment (JohelEGP) [reply source/ranges.tex:11231]: Perhaps that's worth a separate editorial issue. My suggestion seems to be the norm, where macros that internally use `\tcode` are not redundantly wrapped in `\tcode`.

File: source/ranges.tex (hunk 4 above)
Comment (JohelEGP) [reply source/ranges.tex:11214]: That's https://github.com/cplusplus/draft/issues/5250.

File: source/ranges.tex (hunk 5 above)
Comment (JohelEGP) [reply source/ranges.tex:12087]: I meant not an editorial issue for my suggestion here, but for adding `\tcode` everywhere within comments, even if already implied. I leave that to you.

**Pull request merged**